### PR TITLE
[ENH] Add RPC function to delete a response using its ID

### PR DESCRIPTION
**Pull request by arnaudbore**
Add RPC function to delete a response depending on the ID of the response. @olleharstedt 
This PR followed this original one #78

### CODE REVIEW

File: application/helpers/remotecontrol/remotecontrol_handle.php [hunk 1]
```diff
@@ -2749,6 +2749,83 @@ public function update_response($sSessionKey, $iSurveyID, $aResponseData)
...
+    		    $oSurvey = Survey::model()->findByPk($iSurveyID);
+    		    if (!isset($oSurvey)){
+    			      return array('status' => 'Error: Invalid survey ID');
+            }
+
+            if (hasSurveyPermission($iSurveyID, 'responses', 'delete')){
+                // get response id from response table using token
+                $oResult = Survey_dynamic::model($iSurveyID)->findByPk($iResponseID);
+                if ($oResult){
+                    // delete the files
```
Comment (Shnoulle) [inline application/helpers/remotecontrol/remotecontrol_handle.php:2779]: Better to use `Response::model($iSurveyID)->findByPk($iResponseID);` delete file is included 
https://github.com/LimeSurvey/LimeSurvey/blob/f5cbc5748fb8d759ab8fff2fa3b50cc5f3dc6ed9/application/models/Response.php#L10

Then : no need to copy code here.

File: application/helpers/remotecontrol/remotecontrol_handle.php [hunk 2]
```diff
@@ -2749,6 +2749,83 @@ public function update_response($sSessionKey, $iSurveyID, $aResponseData)
...
+                                }
+                            }
+                        }
+                    }
+
+                    // delete the row
+                    Survey_dynamic::model($iSurveyID)->deleteByPk($iResponseID);
+                    // delete timings if savetimings is set
+                    if (isset( $thissurvey['savetimings'] ) && $thissurvey['savetimings'] == "Y"){
+                        Survey_timings::model($iSurveyID)->deleteByPk($iResponseID);
```
Comment (Shnoulle) [inline application/helpers/remotecontrol/remotecontrol_handle.php:2810]: Must add in Reponse model too then. 
https://github.com/LimeSurvey/LimeSurvey/blob/f5cbc5748fb8d759ab8fff2fa3b50cc5f3dc6ed9/application/models/Response.php#L10

**Review by Shnoulle — CHANGES_REQUESTED** [review]
No code only in remote control : remote control must do exactly the same than GUI : not more, not less.

File: application/helpers/remotecontrol/remotecontrol_handle.php (hunk 1 above)
Comment (Shnoulle) [reply application/helpers/remotecontrol/remotecontrol_handle.php:2779]: PS : maybe `SurveyDynamic::model($iSurveyID)->findByPk($iResponseID)->delete();` must call `Response::model($iSurveyID)->findByPk($iResponseID)->delete();` ;)

**Comment by olleharstedt** [comment]
Thanks for the review, Denis!!

**Review by Shnoulle — CHANGES_REQUESTED** [review]
Still some issue, and i think tou can no t use `Survey_dynamic` ? You test if it work ?

**Comment by Shnoulle** [comment]
> Thanks for the review, Denis!!

OK with you too ? Delete timing in response is great for GUI too (needed but no bug report)

File: application/helpers/remotecontrol/remotecontrol_handle.php [hunk 3]
```diff
@@ -2749,6 +2749,55 @@ public function update_response($sSessionKey, $iSurveyID, $aResponseData)
...
+    		    if (!isset($oSurvey)){
+    			      return array('status' => 'Error: Invalid survey ID');
+            }
+
+            if (hasSurveyPermission($iSurveyID, 'responses', 'delete')){
+                // get response id from response table using token
+                $oResult = Response::model($iSurveyID)->findByPk($iResponseID);
+                if ($oResult){
+                    // delete the files
+                    $oResult.deleteFiles();
```
Comment (Shnoulle) [inline application/helpers/remotecontrol/remotecontrol_handle.php:2780]: `$oResult->deleteFiles();` but not needed

File: application/helpers/remotecontrol/remotecontrol_handle.php [hunk 4]
```diff
@@ -2749,6 +2749,55 @@ public function update_response($sSessionKey, $iSurveyID, $aResponseData)
...
+                    // delete timings
+                    $oResult.deleteTimings();
```
Comment (Shnoulle) [inline application/helpers/remotecontrol/remotecontrol_handle.php:2782]: `$oResult->deleteTimings();` but not needed

File: application/helpers/remotecontrol/remotecontrol_handle.php [hunk 5]
```diff
@@ -2749,6 +2749,55 @@ public function update_response($sSessionKey, $iSurveyID, $aResponseData)
...
+    	  // check sessionKey is valid or not
+        if ($this->_checkSessionKey($sSessionKey)){
+    		    $oSurvey = Survey::model()->findByPk($iSurveyID);
+    		    if (!isset($oSurvey)){
+    			      return array('status' => 'Error: Invalid survey ID');
+            }
+
+            if (hasSurveyPermission($iSurveyID, 'responses', 'delete')){
+                // get response id from response table using token
+                $oResult = Response::model($iSurveyID)->findByPk($iResponseID);
```
Comment (Shnoulle) [inline application/helpers/remotecontrol/remotecontrol_handle.php:2777]: No clean rules but `$Response = Response::model($iSurveyID)->findByPk($iResponseID);` seems clearer

https://manual.limesurvey.org/Coding_guidelines#Naming_convention

**Review by Shnoulle — CHANGES_REQUESTED** [review]

**Review by Shnoulle — COMMENTED** [review]
Seems cool

File: application/helpers/remotecontrol/remotecontrol_handle.php [hunk 6]
```diff
@@ -2749,6 +2749,51 @@ public function update_response($sSessionKey, $iSurveyID, $aResponseData)
...
+    			      return array('status' => 'Error: Invalid survey ID');
+            }
+
+            if (hasSurveyPermission($iSurveyID, 'responses', 'delete')){
+                // get response id from response table using ID
+                $Response = Response::model($iSurveyID)->findByPk($iResponseID);
+                if ($Response){
+                    // delete the files and timings and row
+                    $Response->delete();
+                    return array('iSurveyID'=>$iSurveyID);
```
Comment (Shnoulle) [inline application/helpers/remotecontrol/remotecontrol_handle.php:2781]: See 
https://github.com/LimeSurvey/LimeSurvey/blob/f5cbc5748fb8d759ab8fff2fa3b50cc5f3dc6ed9/application/helpers/remotecontrol/remotecontrol_handle.php#L1939 

Maybe `return array($iResponseID=>'deleted');`

**Review by Shnoulle — APPROVED** [review]
Seems Ok, maybe update the return value ? And need to test

**Comment by Shnoulle** [comment]
@arnaudbore i say maybe  : need discussion ;) . You think same ?

**Comment by arnaudbore** [comment]
Sorry, I really thought it was a good idea and decided to move forward.

**Comment by olleharstedt** [comment]
Now all we need is a written test...

**Comment by arnaudbore** [comment]
I don't find any place in the repo where others tests are done. If you tell me where it is I can write the test.

**Comment by olleharstedt** [comment]
Have a look in this folder: https://github.com/LimeSurvey/LimeSurvey/tree/master/tests/unit/helpers/remotecontrol

I think you could copy-paste a lot. Ask if you have any questions.

**Comment by arnaudbore** [comment]
It's in the master not in the 3.x branch.

**Comment by olleharstedt** [comment]
We prefer not to put new features into an old LTS version. Is it OK if we include it in LS 5.1.0 instead?

**Comment by arnaudbore** [comment]
I don't think it will break anything since it's very specific and controlled. The admin won't upgrade to 5.1.0 :( 
We can test it on 5.1.0 and "easily" admit it will have the same behaviour on 3.x. I think this kind of use is very niche. What do you think ?

**Comment by olleharstedt** [comment]
Yep, it's true that the risk is low when adding new API calls, but it should go into 5.1 dev branch first and then be cherry-picked into LTS after.

**Comment by arnaudbore** [comment]
Ok I'll check if a rebase to 5.1 is painless and then add a test.

**Comment by olleharstedt** [comment]
Thanks!!

**Comment by olleharstedt** [comment]
The major difference between LTS and LS4/LS5 is how translations are handled in the database.

**Comment by arnaudbore** [comment]
Which branch/tag should I use ? master ? development ? @olleharstedt

**Comment by olleharstedt** [comment]
Moved to develop branch, but lots of conflicts. Hm. Doesn't make sense. Maybe open a new PR to develop branch directly?

**Comment by arnaudbore** [comment]
Done see #1928

**Pull request closed without merging**